src/test/connect-test.c

```diff
@@ -0,0 +1,444 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+/***
+  This file is part of systemd.
+
+  Copyright 2017 Intel Corporation.
+
+  systemd is free software; you can redistribute it and/or modify it
+  under the terms of the GNU Lesser General Public License as published by
+  the Free Software Foundation; either version 2.1 of the License, or
+  (at your option) any later version.
+
+  systemd is distributed in the hope that it will be useful, but
+  WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public License
+  along with systemd; If not, see <http://www.gnu.org/licenses/>.
+***/
+
+#include <string.h>
+#include <unistd.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <fcntl.h>
+
+#include <arpa/inet.h>
+#include <sys/socket.h>
+#include <sys/types.h>
+#include <netinet/in.h>
+
+#include "sd-event.h"
+#include "time-util.h"
+#include "string-util.h"
+#include "parse-util.h"
+#include "socket-util.h"
+
+#define INPUT_CONNECTION_SUCCESS  "success"
+#define INPUT_CONNECTION_REJECTED "rejected"
+#define INPUT_CONNECTION_TIMEOUT  "timeout"
+
+enum connection_status {
+        CONNECTION_UNDEFINED = 0,
+        CONNECTION_TIMEOUT,
+        CONNECTION_REFUSED,
+        CONNECTION_SUCCEEDED,
+};
+
+struct context {
+        enum connection_status ipv4;
+        enum connection_status ipv6;
+        enum connection_status timeout;
+};
+
+static const char *enum_to_s(enum connection_status status) {
+        switch (status) {
+        case CONNECTION_TIMEOUT:
+                return INPUT_CONNECTION_TIMEOUT;
+        case CONNECTION_REFUSED:
+                return INPUT_CONNECTION_REJECTED;
+        case CONNECTION_SUCCEEDED:
+                return INPUT_CONNECTION_SUCCESS;
+        default:
+                return "undefined";
+        }
+}
+
+static int timeout_handler(sd_event_source *event_source, usec_t usec, void *data) {
+        struct context *ctx = data;
+
+        ctx->timeout = CONNECTION_TIMEOUT;
+
+        sd_event_exit(sd_event_source_get_event(event_source), 0);
+        return 0;
+}
+
+static int server_handler(sd_event_source *event_source, int fd, uint32_t revents, void *data) {
+        int conn_fd, r;
+        union sockaddr_union addr = { 0 };
+        struct sockaddr *sa, addr_name = { 0 };
+        socklen_t addr_length = sizeof(addr), len;
+
+        r = getsockname(fd, &addr_name, &addr_length);
```

Comment:
this will corrupt your memory, as addr_name is too short for storing an IPv4 or IPv6 sockaddr...

please stick to using sockaddr_union, and if you need it as plain sockaddr, then use it's "sa" field:

```
r = getsockname(fd, &addr.sa, &addr_length);
```